File: frontend/src/js/cost.js
// ─── Cost / Quota Monitor ──────────────────────────────────────────────

async function renderCost(content) {
  content.innerHTML = skeleton(4);
  try {
    const [overview, models, oci] = await Promise.all([
      api('/api/cost/overview'),
      api('/api/cost/models'),
      api('/api/cost/oci'),
    ]);

    let html = pageHead('Cost', 'Usage, quotas & free-tier limits', [
      { label: 'Containers', value: overview.container_count },
    ]);

    // OCI free tier status
    const arm = oci['gh-arm'];
    html += '<div class="section"><div class="section-title">OCI Free Tier (gh-arm)</div>';
    html += `<div class="gauge-row">
      ${gauge('Disk', arm.disk_pct)}
      ${gauge('CPU', arm.cpu_pct)}
      ${gauge('Memory', arm.mem_pct)}
    </div>`;
    html += `<div class="section-desc">${arm.compute_limit}</div>`;
    html += '</div>';

    // Network
    html += '<div class="section"><div class="section-title">Network (all hosts)</div>';
    html += `<div class="stat-grid">
      <div class="stat-tile">
        <div class="stat-label">Download</div>
        <div class="stat-value">${overview.network.rx_mbps} <small>Mbps</small></div>
      </div>
      <div class="stat-tile">
        <div class="stat-label">Upload</div>
        <div class="stat-value">${overview.network.tx_mbps} <small>Mbps</small></div>
      </div>
    </div>`;
    html += '</div>';

    // Host resource usage
    html += '<div class="section"><div class="section-title">Host CPU / Memory</div>';
    html += '<div class="card">';
    for (const h of overview.host_cpu) {
      const ip = h.instance.split(':')[0];
      const memEntry = overview.host_mem.find(m => m.instance === h.instance);
      const memPct = memEntry ? Math.round(parseFloat(memEntry.value)) : 0;
      const cpuPct = Math.round(h.cpu_pct);
      html += `
        <div class="list-item">
          <div class="info">
            <div class="title mono-val" style="font-size:13px">${ip}</div>
            <div class="subtitle">CPU <span class="mono-val" style="color:${pctColor(cpuPct)}">${cpuPct}%</span> · MEM <span class="mono-val" style="color:${pctColor(memPct)}">${memPct}%</span></div>
          </div>
        </div>
      `;
    }
    html += '</div></div>';

    // Model usage
    if (models.usage && models.usage.length > 0) {
      html += '<div class="section"><div class="section-title">Model Usage (24h)</div>';
      html += '<div class="card">';
      for (const u of models.usage) {
        html += `
          <div class="list-item">
            <div class="info">
              <div class="title">${u.model}</div>
              <div class="subtitle">${u.provider} · ${u.requests} requests · ${(u.tokens / 1000).toFixed(1)}K tokens</div>
            </div>
          </div>
        `;
      }
      html += '</div></div>';
    }

    // Top containers by memory
    if (overview.containers.length > 0) {
      html += '<div class="section"><div class="section-title">Top Containers (memory)</div>';
      html += '<div class="card">';
      for (const c of overview.containers.slice(0, 10)) {
        html += `<div class="list-item"><div class="info"><div class="title">${c.name}</div></div><div class="mono-val" style="color:var(--hint)">${c.memory_mb} MB</div></div>`;
      }
      html += '</div></div>';
    }

    content.innerHTML = html;
  } catch (err) {
    content.innerHTML = errorCard(err.message);
  }
}

File: backend/app/api/cost.py
"""Cost / Quota Monitor — track API usage, model costs, and resource quotas."""

import asyncio
import httpx
import os
import glob
from fastapi import APIRouter, Depends
from ..core.auth_secure import get_authenticated_user
from ..core.config import PROMETHEUS_URL, SESSION_DB

# Instance-label match for the host tracked by the OCI free-tier endpoint below.
# Set via env var — no real instance/IP fragment ships as a default.
OCI_INSTANCE_MATCH = os.environ.get("OCI_INSTANCE_MATCH", ".*oci-host.*")

router = APIRouter()


async def prom_query_all(query: str) -> list:
    async with httpx.AsyncClient(timeout=10) as c:
        r = await c.get(f"{PROMETHEUS_URL}/api/v1/query", params={"query": query})
        r.raise_for_status()
        data = r.json()
        if data["status"] == "success":
            return data["data"]["result"]
    return []


async def prom_query(query: str) -> str:
    async with httpx.AsyncClient(timeout=10) as c:
        r = await c.get(f"{PROMETHEUS_URL}/api/v1/query", params={"query": query})
        r.raise_for_status()
        data = r.json()
        if data["status"] == "success" and data["data"]["result"]:
            return data["data"]["result"][0]["value"][1]
    return "0"


@router.get("/overview")
async def cost_overview(user = Depends(get_authenticated_user)):
    """Get cost/quota overview."""

    # Container resource usage
    containers = await prom_query_all(
        'sum(container_memory_usage_bytes{name!=""}) by (name)'
    )

    # Per-host resource summary
    host_cpu = await prom_query_all(
        '100 - (avg by(instance) (rate(node_cpu_seconds_total{job="node_exporter",mode="idle"}[5m])) * 100)'
    )
    host_mem = await prom_query_all(
        "100 * (1 - (node_memory_MemAvailable_bytes / node_memory_MemTotal_bytes))"
    )

    # Docker container count
    container_count = await prom_query(
        'count(count(container_last_seen{name!=""}) by (name))'
    )

    # Network traffic (rough daily estimate from rate)
    net_rx = await prom_query("sum(rate(node_network_receive_bytes_total[1h]))")
    net_tx = await prom_query("sum(rate(node_network_transmit_bytes_total[1h]))")

    return {
        "containers": [
            {
                "name": c["metric"].get("name", "?"),
                "memory_mb": round(float(c["value"][1]) / 1e6, 1),
            }
            for c in containers[:20]
        ],
        "host_cpu": [
            {
                "instance": h["metric"].get("instance", "?"),
                "cpu_pct": round(float(h["value"][1]), 1),
            }
            for h in host_cpu
        ],
        "host_mem": [
            {
                "instance": h["metric"].get("instance", "?"),
                "mem_pct": round(float(h["value"][1]), 1),
            }
            for h in host_mem
        ],
        "container_count": int(float(container_count)),
        "network": {
            "rx_mbps": round(float(net_rx) * 8 / 1e6, 2),
            "tx_mbps": round(float(net_tx) * 8 / 1e6, 2),
        },
    }


@router.get("/models")
async def model_usage(user = Depends(get_authenticated_user)):
    """Get Hermes model usage stats from session logs."""
    # Read Hermes session DB for recent usage
    import sqlite3
    import json
    from datetime import datetime, timedelta

    db_path = SESSION_DB
    if not os.path.exists(db_path):
        return {"usage": [], "note": "Session DB not found"}

    try:
        conn = sqlite3.connect(db_path)
        conn.row_factory = sqlite3.Row
        # Try to get recent model usage
        cutoff = (datetime.now() - timedelta(hours=24)).isoformat()
        rows = conn.execute(
            "SELECT model, provider, COUNT(*) as count, "
            "SUM(total_tokens) as tokens "
            "FROM messages WHERE created_at > ? "
            "GROUP BY model ORDER BY count DESC LIMIT 20",
            (cutoff,),
        ).fetchall()
        conn.close()

        return {
            "usage": [
                {
                    "model": r["model"],
                    "provider": r["provider"],
                    "requests": r["count"],
                    "tokens": r["tokens"] or 0,
                }
                for r in rows
            ],
            "period": "24h",
        }
    except Exception as e:
        return {"usage": [], "error": str(e)}


@router.get("/oci")
async def oci_resources(user = Depends(get_authenticated_user)):
    """Get OCI free tier resource usage for the configured host."""
    # Disk is the constrained resource on OCI free tier
    disk = await prom_query(
        f'100 * (1 - (node_filesystem_avail_bytes{{instance=~"{OCI_INSTANCE_MATCH}",mountpoint="/"}} / node_filesystem_size_bytes{{instance=~"{OCI_INSTANCE_MATCH}",mountpoint="/"}}))'
    )
    cpu = await prom_query(
        f'100 - (avg(rate(node_cpu_seconds_total{{instance=~"{OCI_INSTANCE_MATCH}",mode="idle"}}[5m])) * 100)'
    )
    mem = await prom_query(
        f'100 * (1 - (node_memory_MemAvailable_bytes{{instance=~"{OCI_INSTANCE_MATCH}"}} / node_memory_MemTotal_bytes{{instance=~"{OCI_INSTANCE_MATCH}"}}))'
    )

    return {
        "oci-host": {
            "disk_pct": round(float(disk), 1),
            "cpu_pct": round(float(cpu), 1),
            "mem_pct": round(float(mem), 1),
            "disk_limit_gb": 200,
            "compute_limit": "4 OCPU / 24GB RAM (free tier)",
        },
        "note": "OCI free tier: 2×AMD Micro + 1×A1.Flex (4 OCPU/24GB). Storage capped at 200GB.",
    }
